src/analysis_and_optimization/Partial_evaluator.ml

```diff
@@ -787,11 +787,41 @@ let rec eval_expr ?(preserve_stability = false) (e : Expr.Typed.t) =
                   , [ { pattern=
                           FunApp (StanLib ("Times__", FnPlain, mem), [x; y])
                       ; _ }; z ] )
-                 |( "Plus__"
+                  when (not preserve_stability)
+                       && not
+                            ( UnsizedType.is_eigen_type x.meta.type_
+                            && UnsizedType.is_eigen_type y.meta.type_ ) ->
+                    let lub_mem = lub_mem_pat [mem] in
+                    FunApp (StanLib ("fma", suffix, lub_mem), [x; y; z])
+                | ( "Plus__"
                   , [ z
                     ; { pattern=
                           FunApp (StanLib ("Times__", FnPlain, mem), [x; y])
                       ; _ } ] )
+                  when (not preserve_stability)
+                       && not
+                            ( UnsizedType.is_eigen_type x.meta.type_
+                            && UnsizedType.is_eigen_type y.meta.type_ ) ->
+                    let lub_mem = lub_mem_pat [mem] in
+                    FunApp (StanLib ("fma", suffix, lub_mem), [x; y; z])
```

Comment:
These first two cases can stay together, right?

Reply:
They actually have to be broken up, at least when I did this I get a 

```
Error (warning 57): Ambiguous or-pattern variables under guard;
variables x,y may match different arguments. (See manual section 9.5) ocamllsp
```

Reply:
Huh, today I learned. There's a great explanation of why this is true in the manual: https://ocaml.org/releases/4.12/htmlman/comp.html#ss%3Awarn57